include/flexflow/model.h

```diff
@@ -615,22 +634,24 @@ class FFModel {
   // Internal PCG::Node creation APIs
   // ========================================
   template <typename T>
-  PCG::Node get_or_create_node(ParallelTensor const input, typename T::Params const &params) {
-    using Params = typename T::Params;  
+  PCG::Node get_or_create_node(const typename T::Input& input, typename T::Params const &params) {
+    using Params = typename T::Params;
+
+    auto input_shapes = get_input_shape<typename T::Input>(input);
 
-    if (!params.is_valid(input->get_shape())) {
+    if (!params.is_valid(input_shapes)) {
       return PCG::Node::INVALID_NODE;
     }
 
     T *op = nullptr;
 
-    std::pair<ParallelTensorShape, Params> key{input->get_shape(), params};
+    std::pair<typename ToShape<typename T::Input>::type, Params> key{input_shapes, params};
     auto &cache = this->get_cache<T>();
     const auto &it = cache.find(key);
     if (it != cache.end()) {
       op = it->second;
     } else {
-      op = new T(*this, params, input, false, NULL);
+      op = new T(*this, params, input, false/*allocate_weights or inplace_a*/, NULL/*name*/);
```

Comment:
Is this `false` serving as the input to two different parameters (i.e. `allocate_weights` _and_ `inplace_a`) depending on `T`? In that case I think you were right before and we should just have a constructor that defaults this parameter to false.